### ISSUE TITLE
Non-linear growth in processing time with mitmproxy regarding packet size

### ISSUE THREAD

**Issue by jackfromeast** [mitmproxy/mitmproxy#6944]
#### Problem Description
Recently, I've observed an additional overhead that impacts performance which is primarily caused by the `mitmproxy` proxy server. Through my analysis, I identified two critical issues:

1. Non-linear Growth in Processing Time: The processing time for network requests increases non-linearly with the size of the packets. This means that as the packet size grows, the time required to process these requests increases disproportionately.

2. Significant Slowdown with Concurrent Requests: The processing time for network requests slows down dramatically when handling multiple requests simultaneously.

The worst-case scenario occurs when there is a high volume of concurrent requests, especially when some of these requests are significantly large. In such cases, the performance degradation is severe, leading to substantial delays and inefficiencies in processing.

I demonstrated these findings through two experimental settings: 1/ **Sequential Requests**: The first HTML page loads the JavaScript files one by one. Each request is triggered by the `onload` event of the previously requested JavaScript file. 2/ **Concurrent Requests**: The second HTML request loads all the JavaScript files simultaneously. For these experiments, I created 20 JavaScript files ranging from 1MB to 100MB, increasing by 5MB each. I host the proxy and test server on the `localhost` with different ports.

#### Sequential Requests: Without mitmproxy vs. With mitmproxy

![mitmproxy_performance_impact-1](https://github.com/mitmproxy/mitmproxy/assets/51242003/41ae3a87-a3e7-47b8-906c-99a090a41d92)

| Name                | Size     | Loading Time (Without mitmproxy) | Loading Time (With mitmproxy) |
|---------------------|----------|----------------------------------|-------------------------------|
| large_script_1.js   | 1.0 MB   | 4 ms                             | 71 ms                         |
| large_script_2.js   | 6.3 MB   | 35 ms                            | 62 ms                         |
| large_script_3.js   | 11.5 MB  | 83 ms                            | 195 ms                        |
| large_script_4.js   | 16.8 MB  | 142 ms                           | 272 ms                        |
| large_script_5.js   | 22.0 MB  | 201 ms                           | 416 ms                        |
| large_script_6.js   | 27.3 MB  | 254 ms                           | 602 ms                        |
| large_script_7.js   | 32.5 MB  | 294 ms                           | 818 ms                        |
| large_script_8.js   | 37.7 MB  | 357 ms                           | 1.78 s                        |
| large_script_9.js   | 43.0 MB  | 412 ms                           | 2.94 s                        |
| large_script_10.js  | 48.2 MB  | 467 ms                           | 4.34 s                        |
| large_script_11.js  | 53.5 MB  | 502 ms                           | 5.82 s                        |
| large_script_12.js  | 58.7 MB  | 578 ms                           | 7.69 s                        |
| large_script_13.js  | 64.0 MB  | 590 ms                           | 9.29 s                        |
| large_script_14.js  | 69.2 MB  | 664 ms                           | 11.30 s                       |
| large_script_15.js  | 74.4 MB  | 744 ms                           | 13.83 s                       |
| large_script_16.js  | 79.7 MB  | 779 ms                           | 16.19 s                       |
| large_script_17.js  | 84.9 MB  | 832 ms                           | 18.74 s                       |
| large_script_18.js  | 90.2 MB  | 883 ms                           | 20.94 s                       |
| large_script_19.js  | 95.4 MB  | 976 ms                           | 24.02 s                       |
| large_script_20.js  | 101 MB   | 970 ms                           | 26.58 s                       |

**Without mitmproxy:**

![mitm-nproxy-one-by-one](https://github.com/mitmproxy/mitmproxy/assets/51242003/d41b6bc3-c43e-4a39-87d7-83b728f438c8)

**With mitmproxy:**

![mitm-proxy-one-by-one](https://github.com/mitmproxy/mitmproxy/assets/51242003/88f27944-ea83-44c4-9ca4-83c7837c0fb5)

#### Concurrent Requests: Without mitmproxy vs. With mitmproxy

![mitmproxy_performance_impact-2](https://github.com/mitmproxy/mitmproxy/assets/51242003/2060a008-d644-405c-afcc-e89eca323293)

| Name                | Size     | Loading Time (Without mitmproxy) | Loading Time (With mitmproxy) |
|---------------------|----------|----------------------------------|-------------------------------|
| large_script_1.js   | 1.0 MB   | 4 ms                             | 21 ms                         |
| large_script_2.js   | 6.3 MB   | 40 ms                            | 281 ms                        |
| large_script_3.js   | 11.5 MB  | 97 ms                            | 700 ms                        |
| large_script_4.js   | 16.8 MB  | 152 ms                           | 1.36 s                        |
| large_script_5.js   | 22.0 MB  | 202 ms                           | 1.72 s                        |
| large_script_6.js   | 27.3 MB  | 266 ms                           | 2.53 s                        |
| large_script_7.js   | 32.5 MB  | 332 ms                           | 3.53 s                        |
| large_script_8.js   | 37.7 MB  | 435 ms                           | 6.03 s                        |
| large_script_9.js   | 43.0 MB  | 547 ms                           | 9.56 s                        |
| large_script_10.js  | 48.2 MB  | 719 ms                           | 15.74 s                       |
| large_script_11.js  | 53.5 MB  | 940 ms                           | 20.72 s                       |
| large_script_12.js  | 58.7 MB  | 1.05 s                           | 28.10 s                       |
| large_script_13.js  | 64.0 MB  | 1.17 s                           | 36.80 s                       |
| large_script_14.js  | 69.2 MB  | 1.27 s                           | 50.31 s                       |
| large_script_15.js  | 74.4 MB  | 1.64 s                           | 1.1 min                       |
| large_script_16.js  | 79.7 MB  | 2.10 s                           | 1.4 min                       |
| large_script_17.js  | 84.9 MB  | 2.41 s                           | 1.6 min                       |
| large_script_18.js  | 90.2 MB  | 2.54 s                           | 1.9 min                       |
| large_script_19.js  | 95.4 MB  | 2.77 s                           | 2.2 min                       |
| large_script_20.js  | 101 MB   | 2.94 s                           | 2.4 min                       |

**Without mitmproxy:**

![mitm-nproxy-concurrent](https://github.com/mitmproxy/mitmproxy/assets/51242003/11a6b9c8-cc92-45be-9d7e-bac4464bcea2)

**With mitmproxy:**

![mitm-proxy-concurrent](https://github.com/mitmproxy/mitmproxy/assets/51242003/d134c140-c36b-4254-a4c6-94b77f434c36)

#### System Information

```
Mitmproxy: 10.3.0
Python:    3.11.6
OpenSSL:   OpenSSL 3.2.2 4 Jun 2024
Platform:  Linux-6.5.0-28-generic-x86_64-with-glibc2.38
```

**Therefore, I was wondering if there is a way to mitigate this situation that makes mitmproxy good at processing ~100MB JavaScript files if there is no resource limitation? Thanks!**

**Comment by mhils** [comment]
Can you reproduce this on 10.3.1?

**Comment by jackfromeast** [comment]
Yeah, I can reproduce this problem on 10.3.1. Do you have any idea what I can do to fix this issue?

I also attach the script that I used to set up the test server. After running the script, you can simply use `http-server . -p 8081` to host the test server.

```
import os

# Define the initial size and the increment size (in bytes)
initial_size = 1 * 1024 * 1024  # 1MB
increment_size = 5 * 1024 * 1024  # 5MB

# Content to fill the JS files
function_content = "function dummyFunction() { console.log('Hello World!'); }\n"
function_size = len(function_content)

# Create JavaScript files
for i in range(20):
    size = initial_size + i * increment_size
    file_path = f"large_script_{i+1}.js"
    with open(file_path, "w") as f:
        num_functions = size // function_size
        remaining_size = size % function_size
        f.write(function_content * num_functions)
        if remaining_size > 0:
            f.write(function_content[:remaining_size])
    print(f"{file_path} created with size {os.path.getsize(file_path)} bytes")

html_content_onebyone = """
<!DOCTYPE html>
<html>
<head>
    <title>Large JS File Test</title>
</head>
<body>
    <h1>Testing Multiple JavaScript Files</h1>
    <div id="scripts-container"></div>
    <script>
        const scripts = [
"""

# Add the script tags to the HTML content
for i in range(20):
    html_content_onebyone += f'"large_script_{i+1}.js",\n'

html_content_onebyone += """
        ];

        function loadScript(index) {
            if (index >= scripts.length) return;
            const script = document.createElement('script');
            script.src = scripts[index];
            script.onload = () => loadScript(index + 1);
            document.getElementById('scripts-container').appendChild(script);
        }

        loadScript(0);
    </script>
</body>
</html>
"""

# Create the HTML file to include all JavaScript files
html_content_concurrent = """
<!DOCTYPE html>
<html>
<head>
    <title>Large JS File Test</title>
</head>
<body>
    <h1>Testing Multiple JavaScript Files</h1>
    <div id="scripts-container"></div>
"""

# Add the script file paths to the HTML content
for i in range(20):
    html_content_concurrent += f'<script src="large_script_{i+1}.js"></script>\n'

html_content_concurrent += """
</body>
</html>
"""

html_onebyone_path = "index1.html"
html_concurrent_path = "index2.html"

with open(html_onebyone_path, "w") as f:
    f.write(html_content_onebyone)

with open(html_concurrent_path, "w") as f:
    f.write(html_content_concurrent)

print(f"{html_onebyone_path} created and includes all large JavaScript files one by one")
print(f"{html_concurrent_path} created and includes all large JavaScript files")
```

**Comment by mhils** [comment]
> Do you have any idea what I can do to fix this issue?

This looks suspiciously similar to https://github.com/mitmproxy/mitmproxy/issues/6873, but the fix for that made it into 10.3.1. Can you reproduce it with both HTTP/1 and HTTP/2?

**Comment by jackfromeast** [comment]
I could reproduce the issue on HTTP/1.1 by starting Chrome with the '--disable-http2' flag and launching `mitmproxy` with the command `./mitmweb --set http2=false --web-open-browser --web-port 8888 --web-host localhost`. I am not sure about HTTP/2 and HTTP/1 actually. Is this the case?

**Comment by mhils** [comment]
HTTP/1 and HTTP/1.1 is pretty much the same in mitmproxy. The interesting question would be whether we can reproduce it with HTTP/2. That could significantly narrow down where we'd need to search.

Alternatively, you could try to do the same profiling as in https://github.com/mitmproxy/mitmproxy/issues/6873 and see if that turns up something.

**Comment by jackfromeast** [comment]
Yeah, I think I am experiencing the same problem on both HTTP/1.1 and HTTP/2.

```
$ curl --http2 -x http://localhost:8080 -w "time_namelookup:  %{time_namelookup}\ntime_connect:  %{time_connect}\ntime_appconnect:  %{time_appconnect}\ntime_pretransfer:  %{time_pretransfer}\ntime_redirect:  %{time_redirect}\ntime_starttransfer:  %{time_starttransfer}\n----------\ntime_total:  %{time_total}\n" -o /dev/null -s http://localhost:8001/large_script_20.js
time_namelookup:  0.000012
time_connect:  0.000054
time_appconnect:  0.000000
time_pretransfer:  0.000069
time_redirect:  0.000000
time_starttransfer:  27.037522
----------
time_total:  27.048445

$curl --http1.1 -x http://localhost:8080 -w "time_namelookup:  %{time_namelookup}\ntime_connect:  %{time_connect}\ntime_appconnect:  %{time_appconnect}\ntime_pretransfer:  %{time_pretransfer}\ntime_redirect:  %{time_redirect}\ntime_starttransfer:  %{time_starttransfer}\n----------\ntime_total:  %{time_total}\n" -o /dev/null -s http://localhost:8001/large_script_20.js
time_namelookup:  0.000013
time_connect:  0.000054
time_appconnect:  0.000000
time_pretransfer:  0.000068
time_redirect:  0.000000
time_starttransfer:  27.071949
----------
time_total:  27.082602
```

**Comment by jackfromeast** [comment]
I can confirm that the issue I encountered is not the same as the one mentioned in #6873. I attempted to directly return 100MB of data from the intercept script and did not encounter the same problem. @mhils, do you have any ideas on where I should look to fix this issue, considering the profile information provided above?

**Comment by mhils** [comment]
Based on the profile you posted, the most obvious optimization seems to be to replace `(request|response)_body_buf` with a `list[bytes]` implementation where we just append each bytes object, and then do `b"".join(buf)` when we've read it entirely. Does that make sense?

**Comment by jackfromeast** [comment]
I agree that using string concatenation used [here](https://github.com/mitmproxy/mitmproxy/blob/b8fd7eb55de84e385b7eb396c6543867ec2962a9/mitmproxy/proxy/layers/http/__init__.py#L457) could be slow. But will that slow that much?

```
import time

num_chunks = 1000
chunk_size = 100  # 100 bytes

data_chunk = b'a' * chunk_size

def concat_buffer():
    buffer = b""
    for _ in range(num_chunks):
        buffer += data_chunk
    return buffer

def list_buffer():
    buffer = []
    for _ in range(num_chunks):
        buffer.append(data_chunk)
    return b"".join(buffer)

start_time = time.time()
concat_result = concat_buffer()
concat_time = time.time() - start_time

start_time = time.time()
list_result = list_buffer()
list_time = time.time() - start_time

# Verify the results are the same
assert concat_result == list_result

# Print the results
print(f"Time taken using repeated string concatenation: {concat_time:.6f} seconds")
print(f"Time taken using list of chunks: {list_time:.6f} seconds")
```

```
Time taken using repeated string concatenation: 0.001596 seconds
Time taken using list of chunks: 0.000021 seconds
```

Overall, the string concatenation is still fast enough I think for 100MB data.

**Comment by mhils** [comment]
Never undererstimate N^2 😉 
```
chunk_size = 65_000   # ~ one TCP packet
num_chunks = 1024 * 1024 * 100 // chunk_size  # 100mb
```

```
Time taken using repeated string concatenation: 27.177535 seconds
Time taken using list of chunks: 0.033003 seconds
```

**Comment by jackfromeast** [comment]
I think I have fixed this problem with pull request #6952. 

![tmp](https://github.com/mitmproxy/mitmproxy/assets/51242003/6fd5b66e-0f24-452d-bf3e-237d4cdc1406)
![tmp3](https://github.com/mitmproxy/mitmproxy/assets/51242003/898cdde9-a8aa-40f4-9d5e-d6c2f21f2a10)

**Comment by mhils** [comment]
Thank you so much, @jackfromeast! 🍰 🚀
